terragraph_planner/los/cylindrical_los_validator.py

```diff
@@ -149,87 +99,104 @@ def _compute_confidence_by_radius(
-        Use grid center as representative point. Compute the distance from the grid
-        center to the 3d los center line to check if it's in the LOS zone.
+        Use grid center as representative point. Compute the distance from the vertical
+        line that goes through grid center to the 3d los center line to check if LOS
+        is blocked.
+
+        The LOS center line is represented as r = a + p * b, and the vertical
+        line is represented as r = c + q * d, where a, b, c, d are 3-D vectors,
+        and p, q are numbers between (0, 1). Since d is vertical, we use d = (0, 0, -1)
+        here, then b * d = (-by, bx, 0)
+
+        The shortest distance is computed by d = (c - a) b * d / (b * d),
+        where * is cross production between vectors.
+
+        However, we also need to check if the intersection point is on the line segement
+        by computing p and q in the simultaneous equation ((a + p * b) - (c + q * d))b = 0,
+        ((a + p * b) - (c + q * d))d = 0.
+
+        Based on q value, we have two different cases:
+        1. q >= 0, which means the intersection point is not higher than the top of grid.
+           a. If p <= 0 pr p >= 1, which means the block is behind the LOS, this function
+              will return fresnel radius
+           b. otherwise return the distance between two lines.
+
+        2. p < 0, which means the intersection point is higher on the grid. In this
+           case, the function returns the distance from grid top to the LOS center line
+           by first determining where the cloest point is and then computing the distance
+           between two points. If the closest point, which is the intersection of the
+           LOS line and its orthogonal line going through grid top, is not between two
+           sites, return fresnel radius.
         """
-        x, y, z = grid
-        altitude1 = none_throws(site1.altitude)
-        # Gets x, y, z of the nearest point
-        if yx_slope is not None:
-            t = (
-                (x - site1.utm_x)
-                + yx_slope * (y - site1.utm_y)
-                + none_throws(zx_slope) * (z - altitude1)
-            ) / sum_slope_sq
-            # If the orthogonal point is not on the line, use the nearest end
-            if site2.utm_x > site1.utm_x:
-                t = min(max(t, 0), site2.utm_x - site1.utm_x)
+        cx, cy, cz = grid
+        p = ((cx - ax) * bx + (cy - ay) * by) / b_len_2d_sq
+        q = cz - az - p * bz
+        if q >= 0:
+            # Compute distance beween line b and line d
+            if 0 <= p <= 1:
+                return ((ax - cx) * by + (cy - ay) * bx) / b_len_2d
             else:
-                t = min(max(t, site2.utm_x - site1.utm_x), 0)
-            x_n = site1.utm_x + t
-            y_n = site1.utm_y + t * yx_slope
-            z_n = altitude1 + t * none_throws(zx_slope)
+                # The obstruction is behind the site
+                return self._fresnel_radius
```

Comment:
It seems that this is a little misleading here. Because you later clamp this down to the min radius, it's ok. But viewed in isolation, it's a little funny. Perhaps returning None here when the closest point is outside the LOS path is a little better. Then at the call site, if you get None, the confidence returned is 1.0.